activity/coordinated_worker.go

```diff
@@ -15,121 +15,103 @@ const (
 	}
+	w.AddHandler(&ActivityHandler{
+		Activity:    handler.Activity,
+		HandlerFunc: adapter.coordinate,
+		Input:       handler.Input,
+	})
+}
 
-	heartbeats := time.Tick(heartbeatInterval)
-
-	handlerFunc := func(activityTask *swf.ActivityTask, input interface{}) {
-		handlerStartFunc(activityTask, input)
-		for {
-			select {
-			case e := <-startErr:
-				log.Printf("workflow-id=%s activity-id=%s activity-id=%s at=start-error error=%q", LS(activityTask.WorkflowExecution.WorkflowID), LS(activityTask.ActivityType.Name), LS(activityTask.ActivityID), e)
-				w.fail(activityTask, e)
-				return
-			case tick := <-ticks:
-				if tick.Continue {
-					handlerTickFunc(activityTask, input)
-					if tick.Result != nil {
-						//send an activity update when the result is not null, but we are continuing
-						err := w.signalUpdate(activityTask, tick.Result)
-						if err != nil {
-							log.Printf("workflow-id=%s activity-id=%s activity-id=%s at=signal-update-error error=%q", LS(activityTask.WorkflowExecution.WorkflowID), LS(activityTask.ActivityType.Name), LS(activityTask.ActivityID), err)
-							w.fail(activityTask, err)
-							log.Printf("workflow-id=%s activity-id=%s activity-id=%s at=signal-update-error-fail-task error=%q", LS(activityTask.WorkflowExecution.WorkflowID), LS(activityTask.ActivityType.Name), LS(activityTask.ActivityID), err)
-							handler.Cancel(activityTask, tick.Result)
-							log.Printf("workflow-id=%s activity-id=%s activity-id=%s at=signal-update-error-cancel-handler error=%q", LS(activityTask.WorkflowExecution.WorkflowID), LS(activityTask.ActivityType.Name), LS(activityTask.ActivityID), err)
+type coordinatedActivityAdapter struct {
+	heartbeatInterval time.Duration
+	tickMinInterval   time.Duration
+	worker            *ActivityWorker
+	handler           *CoordinatedActivityHandler
+}
 
-						} else {
-							log.Printf("workflow-id=%s activity-id=%s activity-id=%s at=signal-update", LS(activityTask.WorkflowExecution.WorkflowID), LS(activityTask.ActivityType.Name), LS(activityTask.ActivityID))
-						}
-					}
-				} else {
-					if tick.Error != nil {
-						w.fail(activityTask, tick.Error)
-					} else {
-						w.result(activityTask, tick.Result)
-					}
+func (c *coordinatedActivityAdapter) heartbeat(activityTask *swf.ActivityTask, stop <-chan struct{}, cancelActivity chan error) {
+	heartbeats := time.NewTicker(c.heartbeatInterval)
+	defer heartbeats.Stop()
+	for {
+		select {
+		case <-heartbeats.C:
+			if status, err := c.worker.SWF.RecordActivityTaskHeartbeat(&swf.RecordActivityTaskHeartbeatInput{
+				TaskToken: activityTask.TaskToken,
+			}); err != nil {
+				if ae, ok := err.(aws.APIError); ok && ae.Code == ErrorTypeUnknownResourceFault && strings.Contains(ae.Message, TaskGone) {
+					log.Printf("workflow-id=%s activity-id=%s activity-id=%s at=activity-gone", LS(activityTask.WorkflowExecution.WorkflowID), LS(activityTask.ActivityType.Name), LS(activityTask.ActivityID))
+					cancelActivity <- nil
```

Comment:
this is true :) it will just try to `RespondActivityTaskFailed` though right? Which will respond with a task gone as well, and everything will be stopped.